draft-irtf-cfrg-hash-to-curve.md

```diff
@@ -1439,159 +1538,248 @@ Steps:
+given below ({{hashtofield-expand}}).
 
 {{domain-separation}} discusses requirements for domain separation and
-recommendations for choosing domain separation tags. The hash\_to\_curve
-function takes such a tag as a parameter, DST; this is the REQUIRED
-method for applying domain separation.
-
-{{hashtobase-impl}} details the hash\_to\_base procedure.
-
-## Performance considerations {#hashtobase-perf}
-
-The hash\_to\_base function uses HKDF-Extract to combine the
-input msg and domain separation tag DST into a short digest, which is then
-passed to HKDF-Expand {{!RFC5869}}.
-For short messages, this entails at most two extra invocations of H, which
-is a negligible overhead in the context of hashing to elliptic curves.
-
-A related issue is that the random oracle construction described in {{roadmap}}
-requires evaluating two independent hash functions H0 and H1 on msg.
-One way to instantiate independent hashes is to append a counter to
-the value being hashed, e.g., H(msg || 0) and H(msg || 1).
-If msg is long, however, this is either inefficient (because it entails hashing
-msg twice) or requires non-black-box use of H (e.g., partial evaluation).
-
-To sidestep both of these issues, hash\_to\_base takes a second argument, ctr,
-which it passes to HKDF-Expand.
-This means that two invocations of hash\_to\_base on the same msg with different
-ctr values both start with identical invocations of HKDF-Extract.
-This is an improvement because it allows sharing one evaluation of HKDF-Extract
-among multiple invocations of hash\_to\_base, i.e., by factoring out the common
-computation.
-
-## Implementation {#hashtobase-impl}
-
-The following procedure implements hash\_to\_base.
+recommendations for choosing DST, the domain separation tag.
+This is the REQUIRED method for applying domain separation.
 
 ~~~
-hash_to_base(msg, ctr)
+hash_to_field(msg, count)
 
 Parameters:
 - DST, a domain separation tag (see discussion above).
-- H, a cryptographic hash function.
 - F, a finite field of characteristic p and order q = p^m.
 - p, the characteristic of F (see immediately above).
 - m, the extension degree of F, m >= 1 (see immediately above).
 - L = ceil((ceil(log2(p)) + k) / 8), where k is the security
   parameter of the cryptosystem (e.g., k = 128).
-- HKDF-Extract and HKDF-Expand are as defined in RFC5869,
-  instantiated with the hash function H.
+- expand_message, a function that expands an octet-string and
+  domain separation tag into a pseudorandom octet-string
+  (see discussion above).
 
 Inputs:
-- msg is the message to hash.
-- ctr is 0, 1, or 2.
-  This is used to efficiently create independent
-  instances of hash_to_base (see discussion above).
+- msg is an octet-string containing the message to hash.
+- count is the number of elements of F to output.
+
+Outputs:
+- (u_1, ..., u_(count)), a list of field elements.
+
+Notation:
+- For an octet-string str, let str[a : b] represent the
```

Comment:
`str[ a : b ]` means the bytes from str[a] to str[b] both included and `b>=a`?

Reply:
It's exclusive. from str[a] to str[b] inclusive is (b - a + 1) characters.

Reply:
To be clearer, it includes str[a] but excludes str[b].

Reply:
I clarified slightly.